### PR TITLE
Feat/job application show

### LINKED ISSUE #33: Job Application Show Page BE
A logged in user should be able to see the saved details of their individual job applications.

Acceptance Criteria:

- [ ] Utilize SAD pathing to handle when company_id does not exist on user

- [ ] Return 200 status code with correct response

- [ ] Return 400 status code with error messages if validations are not met

- [ ] Return 400 level status code with error message for SAD path

Dependencies: User exists in Database, Company exists in Database, Job Application exists in DB, Job Application tied to User and Company
Blockers: Require Schema definition for table relationships

Additional Information:
Request:
``` ruby
PUTS /api/v1/users/user.id/job_applications/job_application.id
```

Example Response body:
``` ruby
{:data=>
  {:id=>"6",
   :type=>"job_application",
   :attributes=>
    {:position_title=>"Jr. CTO",
     :date_applied=>"2024-10-31",
     :status=>1,
     :notes=>"Fingers crossed!",
     :job_description=>"Looking for Turing grad/jr dev to be CTO",
     :application_url=>"www.example.com",
     :contact_information=>"boss@example.com",
     :company_id=>6
}}}
```

### PR DESCRIPTION
**Pull request by kylomite**
## Type of Change
- [x] feature ⛲
- [x] documentation update 📃
- [x] testing 🧪

## Description
- Show user's job application
- Testing for job application Show method
- Updated seed file
- Updated `Readme.md`

## Motivation and Context

- This PR creates an endpoint for viewing a user's specific job applications based on its ID.
- It handles SAD paths by returning an error response if
  - the job application ID does not exist
  - the job application ID belongs to another user 
  - Closes #33  

## Related Tickets
- #6 

## Added Test?
- [x] Yes 🫡
  - request tests for HAPPY and SAD paths in `job_application_show_spec`

- [x] All previous tests still pass 🥳

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: db/seeds.rb
Comment (Sgalvin36) [inline db/seeds.rb:1]: I like that you added a seed of the job application. That way you can use it as like default values for the front end and do some testing in the development and testing stages. And you didn't create a bunch of seeds either which would populate even the production database with a bunch of dummy data.

File: app/controllers/api/v1/job_applications_controller.rb [hunk 1]
```diff
@@ -11,6 +11,17 @@ def create
     end
   end
 
+  def show
+    user = User.find(params[:user_id])
+    job_application = JobApplication.find_by(id: params[:id])
+
+    if job_application.nil? || job_application.user_id != user.id
```
Comment (Sgalvin36) [inline app/controllers/api/v1/job_applications_controller.rb:18]: Would we want a different error if the user doesn't match?

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb
Comment (Sgalvin36) [inline spec/requests/api/v1/job_applications/job_application_show_spec.rb:1]: Another potential edge case would be a lack of parameters being sent.

**Review by Sgalvin36 — APPROVED** [review]
Very nice work! Its clear and easy to work through

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb
Comment (kylomite) [reply spec/requests/api/v1/job_applications/job_application_show_spec.rb:1]: This is worth doing. I will incorporate this edge case in the testing, if it requires reworking the method, I will incorporate that as well. Thank you :)

File: app/controllers/api/v1/job_applications_controller.rb (hunk 1 above)
Comment (kylomite) [reply app/controllers/api/v1/job_applications_controller.rb:18]: Kinda torn on this, Im not sure if the effort is worth the result, but i'm willing to hear out justifications. Feel free to post some in this thread if anyone is interested.

Comment (Sgalvin36) [reply app/controllers/api/v1/job_applications_controller.rb:18]: Not 100% one way or the other, but clearer messages to the front is helpful and since a lack of job application is a slightly different error than an incorrect user id. They both result in the job application not working as intended, so I also see why one error for each does the job.

**Comment by reneemes** [comment]
Excellent job Kyle! I pulled your branch down to my local to run the tests and everything seems to be running smoothly. I also noticed that you have 100% test coverage of your controller and model with SimpleCov.
One thought, if sad path testing for invalid authorization tokens is handled elsewhere in the codebase, it might not need duplication here. Otherwise, it’s a detail that could further strengthen the tests for this endpoint. Just something to consider. I know Shane mentioned testing with missing parameters, and I agree this could be a useful variable to cover. Overall, this is fantastic work!

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb
Comment (kylomite) [reply spec/requests/api/v1/job_applications/job_application_show_spec.rb:1]: Edge case has been addressed and read me has been updated accordingly, If the change is suitable, feel free to resolve conversation and/or resolve conflict to merge :) Reach out if there Is anything I can help clarify

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb [hunk 2]
```diff
@@ -109,6 +109,18 @@
...
+        get "/api/v1/users/#{@user.id}/job_applications/"
+
+        expect(response).to_not be_successful
+        expect(response.status).to eq(400)
+
+        json = JSON.parse(response.body, symbolize_names: true)
+
+        expect(json[:message]).to eq("Job application ID is missing")
+        expect(json[:status]).to eq(400)
+      end
```
Comment (reneemes) [inline spec/requests/api/v1/job_applications/job_application_show_spec.rb:123]: Thank you for adding this test and being very through with your work. I like your error message here, as I feel it is very informative and user-friendly.

File: app/controllers/api/v1/job_applications_controller.rb [hunk 3]
```diff
@@ -12,6 +12,10 @@ def create
   end
 
   def show
+    if params[:id].blank?
+      render json: ErrorSerializer.format_error(ErrorMessage.new("Job application ID is missing", 400)), status: :bad_request
+      return
+    end
```
Comment (reneemes) [inline app/controllers/api/v1/job_applications_controller.rb:18]: This is a nice addition to the show action. I noticed that you are grabbing the ID from the params, and not from the authentication token. I'm curious about this decision, as both the companies and contacts are collecting it from the token.

Comment (reneemes) [reply app/controllers/api/v1/job_applications_controller.rb:18]: I noticed that the JobApplications controller is not utilizing the authentication tokens or the `authenticate_user` method that has been defined in the ApplicationController. Other controllers in the application are using these for authentication, so I wanted to flag this for consistency and security purposes.

Could you clarify if this was intentional or an oversight? If it's an oversight, it might be worth adding the necessary authentication logic to ensure this controller aligns with the others. I'm curious to hear your thoughts!

**Review by reneemes — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/job_applications_controller.rb (hunk 3 above)
Comment (kylomite) [reply app/controllers/api/v1/job_applications_controller.rb:18]: At first I felt the need to address this refactor in this ticket. However, since the integration of gems related to authorization and authentication there has been a new ticket created (ticket #42) that was made for the express purpose of handling these types of refactors, that being said. I think we should merge the PR as it currently stands so it does not serve as a blocker for the person who will pick up the controller refactor ticket.

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,152 @@
+require "rails_helper"
+
+RSpec.describe "Job Application #show", type: :request do
```
Comment (reneemes) [inline spec/requests/api/v1/job_applications/job_application_show_spec.rb:3]: This is very through testing of both happy and sad paths. Excellent work, Kyle!

File: spec/requests/api/v1/job_applications/job_application_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      #   headers: {"Authorization" => "Bearer #{token}" }, as: :json
+
+      #   expect(response).to_not be_successful
+      #   expect(response.status).to eq(400)
+
+      #   json = JSON.parse(response.body, symbolize_names: true)
+
+      #   expect(json[:message]).to eq("Job application ID is missing")
+      #   expect(json[:status]).to eq(400)
+      # end
```
Comment (reneemes) [inline spec/requests/api/v1/job_applications/job_application_show_spec.rb:149]: We have worked together on this sad path test in attempts to get it up and running. I hope we can revisit it in the future, as I believe it’s an important edge case to cover. With the recent addition of authorizations, we encountered an issue related to the Pundit gem and the show action route.
